packages/labs/router/README.md

```diff
@@ -2,6 +2,255 @@
+
+Usage will generally look like this, with a configuration in a reactive controller, and rendering done via route-specific render callbacks and an "outlet" to use in the main render() method:
+
+```ts
+class MyElement extends LitElement {
+  private _routes = new Routes(this, [
+    {path: '/', render: () => html`<h1>Home</h1>`},
+    {path: '/projects', render: () => html`<h1>Projects</h1>`},
+    {path: '/about', render: () => html`<h1>About</h3>`},
+  ]);
+
+  render() {
+    return html`
+      <header>...</header>
+      <main>${this._routes.outlet}</main>
+      <footer>...</footer>
+    `;
+  }
+}
+```
+
+Routes can be nested: a route path can include a trailing `/*` pattern to match against a prefix, and will automatically propagate that prefix to Routes controllers defined in child elements.
+
+The general shape of the API includes:
+
+- A `Router` controller that's used as a top-level singleton to set up event listeners
+- A `Routes` controller for declaring routes inside components
+- Declaration of routes with [`URLPattern`](https://developer.mozilla.org/en-US/docs/Web/API/URLPattern) and render callbacks
+- Extraction of URL pattern parameters into data objects passed to render callbacks
+- A `routes.outlet` that renders the current route's render callback
+- A `routes.link()` method to generate URLs to use in `<a>` tags, etc.
+- A `routes.goto()` method for performing a navigation
+
+## API
+
+### Router
+
+A Router is a controller (a subclass of the Routes controller) for use at the top-level of an application. It's main purpose is to set up global `click` and `popstate` event listeners (which should be installed only once on a page). It can optionally contain route definitions as a convenience.
+
+It can be installed with no configuration:
+
+```ts
+class App extends LitElement {
+  private router = new Router(this);
+}
+```
+
+Or contain route configurations:
+
+```ts
+class MyElement extends LitElement {
+  private router = new Router(this, [
+    {path: '/', render: () => html`<h1>Home</h1>`},
+  ]);
+
+  render() {
+    return router.outlet;
+  }
+}
+```
+
+### Routes
+
+Routes is the main interface into the router API. A Routes controller contains route definitions and the templates that each route renders:
+
+```ts
+class MyElement extends LitElement {
+  private routes = new Routes(this, [
+    {path: '/', render: () => html`<h1>Home</h1>`},
+    {path: '/projects', render: () => html`<h1>Projects</h1>`},
+    {path: '/about', render: () => html`<h1>About</h1>`},
+  ]);
+}
+```
+
+The second argument is the route configuration: an array of `RouteConfig` objects.
+
+#### RouteConfig
+
+A RouteConfig contains at the minimum the pattern to match URLs against and a template to render. Names can be provided to reference routes for link generation.
+
+There are two types of `RouteConfig`s: `PathRouteConfig` and `URLPatternRouteConfig`:
```

Comment:
This is there because if we make URLPattern objects for you, it seems good to let you bring your own. It's only the difference between:

```ts
{
  path: '/foo/:id',
  render({id}) { ... }
}
```

and

```ts
{
  pattern: new URLPattern({pathname: '/foo/:id'}),
  render({id}) { ... }
}
```

The key additional feature that making your own URLPattern allows now is to match on query params. That seems niche enough to relegate to bring-you-own-pattern-object.

Reply:
OK.